File: leetcode/170-two-sum-iii-data-structure-design/main.go
package main

/*
	questions to ask:
	- will there be duplicate number? yes
	- so if total input = [2,2], target=4, result=true? yes
	- so if total input = [2], target=2, no result since we dont count the number on the same index? yes
*/

/*
	1st approach:
	- classic 2 sum but dont count the numbers on the same indeces
	- save the occurence of each number to deal with the duplicate input

	Time	O(n)
	Space	O(n)
	100ms beats 100%
*/

type TwoSum struct {
	Arr []int
	Ht  map[int]int
}

/** Initialize your data structure here. */
func Constructor() TwoSum {
	return TwoSum{[]int{}, make(map[int]int)}
}

/** Add the number to an internal data structure.. */
func (this *TwoSum) Add(number int) {
	this.Arr = append(this.Arr, number)
	if _, x := this.Ht[number]; !x {
		this.Ht[number] = 1
	} else {
		// if the num presents more than once, just save its occurence
		this.Ht[number]++
	}
}

/** Find if there exists any pair of numbers which sum is equal to the value. */
func (this *TwoSum) Find(value int) bool {
	for _, num := range this.Arr {
		target := value - num
		if v, x := this.Ht[target]; x {
			if target != num {
				return true
			} else if v > 1 {
				// it means if there are no than one number in the hashtable
				// there must be 2 numbers on the different indeces that sum up to target
				// case: total=[2,2], target=4
				return true
			}
		}
	}
	return false
}

/**
* Your TwoSum object will be instantiated and called as such:
* obj := Constructor();
* obj.Add(number);
* param_2 := obj.Find(value);
 */

func main() {

}

File: leetcode/170-two-sum-iii-data-structure-design/main.py
"""
    0th: binary search + 2 pointers

    Time of add()   O(N)
    Time of find()  O(N)
    Space           O(N)
    688 ms, faster than 12.52%
"""


from collections import *


class TwoSum:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.nums = []

    def add(self, number: int) -> None:
        """
        Add the number to an internal data structure..
        """
        i = self.__upperBsearch(self.nums, number)
        self.nums.insert(i, number)

    def __upperBsearch(self, nums, target):
        left = 0
        right = len(nums)
        while left < right:
            mid = (left + right) // 2
            if target >= nums[mid]:
                left = mid + 1
            else:
                right = mid
        return left

    def find(self, value: int) -> bool:
        """
        Find if there exists any pair of numbers which sum is equal to the value.
        """
        left = 0
        right = len(self.nums) - 1
        while left < right:
            total = self.nums[left] + self.nums[right]
            if total < value:
                left += 1
            elif total > value:
                right -= 1
            else:
                return True
        return False

# Your TwoSum object will be instantiated and called as such:
# obj = TwoSum()
# obj.add(number)
# param_2 = obj.find(value)


"""
    2nd: hashtable

    Time of add()   O(1)
    Time of find()  O(N)
    Space           O(N)
    336 ms, faster than 69.16%
"""


class TwoSum:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.ht = Counter()

    def add(self, number: int) -> None:
        """
        Add the number to an internal data structure..
        """
        self.ht[number] += 1

    def find(self, value: int) -> bool:
        """
        Find if there exists any pair of numbers which sum is equal to the value.
        """
        for num in self.ht:
            remain = value - num
            if remain in self.ht:
                if num != remain:
                    return True
                elif self.ht[num] > 1:
                    return True
        return False
